launcher/container_runner_test.go

```diff
@@ -0,0 +1,238 @@
+		t.Errorf("Token written to file does not match expected token: got %v, want %v", data, expectedToken)
+	}
+}
+
+func TestTokenIsNotChangedIfRefreshFails(t *testing.T) {
+	ctx := context.Background()
+
+	expectedToken := createJWTToken(t, 5*time.Second)
+	ttl := 5 * time.Second
+	successfulAttestFunc := func(context.Context) ([]byte, error) {
+		return expectedToken, nil
+	}
+
+	errorAttestFunc := func(context.Context) ([]byte, error) {
+		return nil, errors.New("attest unsuccessful")
+	}
+
+	runner := ContainerRunner{
+		attestAgent: &fakeAttestationAgent{attestFunc: successfulAttestFunc},
+	}
+	defer func() {
+		if runner.tokenRefresher.timer != nil {
+			runner.tokenRefresher.timer.Stop()
+			runner.tokenRefresher.done <- true
+		}
+	}()
+
+	if err := runner.fetchAndWriteToken(ctx); err != nil {
+		t.Fatalf("fetchAndWriteToken failed: %v", err)
+	}
+
+	filepath := path.Join(HostTokenPath, attestationVerifierTokenFile)
+	data, err := os.ReadFile(filepath)
+	if err != nil {
+		t.Fatalf("Failed to read from %s: %v", filepath, err)
+	}
+
+	if !bytes.Equal(data, expectedToken) {
+		t.Errorf("Initial token written to file does not match expected token: got %v, want %v", data, expectedToken)
+	}
+
+	// Change attest agent to return error.
+	runner.attestAgent = &fakeAttestationAgent{attestFunc: errorAttestFunc}
+
+	time.Sleep(ttl)
+
+	data, err = os.ReadFile(filepath)
+	if err != nil {
+		t.Fatalf("Failed to read from %s: %v", filepath, err)
+	}
+
+	if !bytes.Equal(data, expectedToken) {
+		t.Errorf("Expected token to remain the same after unsuccessful refresh attempt: got %v", data)
+	}
+}
+
+func TestTokenRefresh(t *testing.T) {
+	ctx := context.Background()
+
+	expectedToken := createJWTToken(t, 5*time.Second)
+
+	ttl := 5 * time.Second
+
+	runner := ContainerRunner{
+		attestAgent: &fakeAttestationAgent{
+			attestFunc: func(context.Context) ([]byte, error) {
+				return expectedToken, nil
+			},
+		},
+	}
+	defer func() {
+		if runner.tokenRefresher.timer != nil {
+			// Drain the timer channel if expired.
+			if !runner.tokenRefresher.timer.Stop() {
+				<-runner.tokenRefresher.timer.C
+			}
+			runner.tokenRefresher.done <- true
+		}
+	}()
+
+	if err := runner.fetchAndWriteToken(ctx); err != nil {
+		t.Fatalf("fetchAndWriteToken failed: %v", err)
+	}
+
+	filepath := path.Join(HostTokenPath, attestationVerifierTokenFile)
+	data, err := os.ReadFile(filepath)
+	if err != nil {
+		t.Fatalf("Failed to read from %s: %v", filepath, err)
+	}
+
+	if !bytes.Equal(data, expectedToken) {
+		t.Errorf("Initial token written to file does not match expected token: got %v, want %v", data, expectedToken)
+	}
+
+	// Change attest agent to return new token.
+	expectedRefreshedToken := createJWTToken(t, 10*time.Second)
+	runner.attestAgent = &fakeAttestationAgent{
+		attestFunc: func(context.Context) ([]byte, error) {
+			return expectedRefreshedToken, nil
+		},
+	}
+
+	time.Sleep(ttl)
+
+	data, err = os.ReadFile(filepath)
```

Comment:
nice test! would it be possible to test that we change the `attestFunc` but don't sleep until the TTL (then assert the data is the old data, not expectedRefreshedToken like in this test)

Reply:
Added a check after changing `attestFunc` but before sleeping, which should cover it.